tools/autograd/templates/ViewFuncs.h

```diff
@@ -0,0 +1,149 @@
+#pragma once
+
+// ${generated_comment}
+
+#include <torch/library.h>
+
+#include <c10/core/SymIntArrayRef.h>
+
+#ifndef AT_PER_OPERATOR_HEADERS
+#include <ATen/Operators.h>
+#else
+$ops_headers
+#endif
+
+namespace torch { namespace autograd {
+
+// forward declaration so it can be a friend class of ViewFunc
+struct ViewFuncSavedStateGuard;
+
+/// Base class for view functions, providing reapplication of a view on a new base.
+/// Each view op should get a codegenerated subclass of this class containing
+/// any state needed to reconstruct the view. The class also provides convenience
+/// accessors for saved SymInts / tensor state. This is useful for e.g. fake-ification,
+/// where we want to use symbolic values or fake tensors instead.
+struct TORCH_API ViewFunc {
+  virtual ~ViewFunc() {}
+  /// Returns any SymInts in the saved state.
+  virtual std::vector<c10::SymInt> get_symints() const = 0;
+  /// Returns any Tensors in the saved state.
+  virtual std::vector<at::Tensor> get_tensors() const = 0;
+  /// Reapplies the view on the given base using the saved state.
+  /// If specified, the provided values for SymInts / Tensors
+  /// are hot-swapped in for the view call.
+  virtual at::Tensor operator()(
+      const at::Tensor&,
+      const std::optional<std::vector<c10::SymInt>>& = c10::nullopt,
```

Comment:
std::optional of vector is problematic, because it says that the optional must hold an OWNING reference to the vector. This means that if you have a vector of SymInt and you want to call this operator, you have to copy the vector *into* a temporary optional, before you can pass a const reference of it to the function. Assuming that it is profitable to pass the vector in by reference (i.e., you aren't going to make a copy of it), it would be better to pass an `optional<SymIntArrayRef>`, which is much more lightweight. Ditto for Tensor (use ArrayRef).

Reply:
I reverted the `operator()` signature and now have `clone_and_set(optional<vector<SymInt>>, optional<vector<Tensor>>)`. For this function, we do expect to copy (or rather move) the provided vector, so it appears to me it's no longer profitable to switch to `optional<ArrayRef>`s, but I'm happy to switch if I'm wrong.